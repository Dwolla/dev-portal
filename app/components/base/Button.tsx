import styled from "@emotion/styled";
import { css } from "@emotion/core";
import {
  WHITE_PRIMARY,
  PURPLE_PRIMARY_BUTTON,
  PURPLE_PRIMARY_HOVER,
  PURPLE_PRIMARY_ACTIVE,
  PINK_SECONDARY_BUTTON,
  PINK_SECONDARY_HOVER,
  PINK_SECONDARY_ACTIVE,
  GREY_4,
  GREY_9,
  GREY_5,
  PURPLE_DARK,
} from "../colors";
import { BOX_SHADOW_6 } from "../shadowDepths";
import { POPPINS } from "../typography";

type Props = {
  text: string;
  size: "tiny" | "small" | "standard" | "large" | "block";
  variant:
    | "primary"
    | "secondary"
    | "hollow-light"
    | "hollow-dark"
    | "hollow-white";
  type?: "button" | "submit" | "reset";
  link?: { href: string; external?: boolean };
};

const Button = ({ text, size, variant, type = "button", link }: Props) =>
  link && link.href ? (
    <ButtonLink text={text} size={size} variant={variant} link={link} />
  ) : (
    <StyledButton type={type} className={`${variant} ${size}`}>
      {text}
    </StyledButton>
  );

const buttonStyles = css`
  color: ${WHITE_PRIMARY};
  font-family: ${POPPINS};
  font-style: thin;
  border-radius: 5px;
  cursor: pointer;
  text-decoration: none;
  :focus {
    outline: 0;
  }
  &.tiny {
    padding: 5px 11px;
    font-size: 11px;
    font-weight: 500;
  }
  &.small {
    padding: 5px 15px;
    font-size: 13px;
  }
  &.standard {
    padding: 10px 15px;
    font-size: 14px;
  }
  &.large {
    padding: 15px 25px;
    font-size: 16px;
  }
  &.block {
    padding: 10px 15px;
    font-size: 14px;
    display: block;
    width: 100%;
    text-align: center;
  }

  &.primary {
    background-color: ${PURPLE_PRIMARY_BUTTON};
    border-color: transparent;

    :hover,
    :focus {
      background-color: ${PURPLE_PRIMARY_HOVER};
      box-shadow: ${BOX_SHADOW_6};
    }

    :active {
      background-color: ${PURPLE_PRIMARY_ACTIVE};
    }
  }
  &.secondary {
    background-color: ${PINK_SECONDARY_BUTTON};
    border-color: transparent;

    :hover,
    :focus {
      background-color: ${PINK_SECONDARY_HOVER};
      box-shadow: ${BOX_SHADOW_6};
    }

    :active {
      background-color: ${PINK_SECONDARY_ACTIVE};
    }
  }
  &.hollow-light {
    border: 1px solid ${GREY_4};
    color: ${GREY_9};
    :hover,
    :focus {
      border-color: ${GREY_5};
    }
  }
  &.hollow-dark {
    border: 1px solid ${WHITE_PRIMARY};
    color: ${WHITE_PRIMARY};
    background: transparent;
    :hover,
    :focus {
      background-color: ${WHITE_PRIMARY};
      color: ${PURPLE_DARK};
    }
  }
  &.hollow-white {
    border: 1px solid transparent;
    background: #ffffff;
    color: #2d2d48;
    :hover,
    :focus {
      border-color: ${GREY_5};
    }
  }
`;

const StyledButton = styled.button`
  ${buttonStyles}
`;

const StyledLink = styled.a`
  ${buttonStyles}
  display: inline-block;
`;

const ButtonLink = ({ text, size, variant, link }: Props) => (
  <StyledLink
    href={link.href}
    target={link.external && "_blank"}
    className={`${variant} ${size}`}
  >
    {text}
  </StyledLink>
);

export default Button;
