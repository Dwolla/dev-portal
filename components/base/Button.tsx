import React from "react";
import styled from "@emotion/styled";

import {
  WHITE_PRIMARY,
  PURPLE_PRIMARY_BUTTON,
  PURPLE_PRIMARY_HOVER,
  PURPLE_PRIMARY_ACTIVE,
  PINK_SECONDARY_BUTTON,
  PINK_SECONDARY_HOVER,
  PINK_SECONDARY_ACTIVE,
  GREY_4,
  GREY_9,
  GREY_5,
  PURPLE_DARK,
} from "../colors";
import { BOX_SHADOW_6 } from "../shadowDepths";
import { POPPINS } from "../typography";

type Props = {
  text: string;
  size: "tiny" | "small" | "standard" | "large";
  variant: "primary" | "secondary" | "hollow-light" | "hollow-dark";
};

function Button({ text, size, variant }: Props) {
  return (
    <ButtonStyle type="button" className={`${size} ${variant}`}>
      <SpanStyle className={`${size}`}>{text}</SpanStyle>
    </ButtonStyle>
  );
}

const ButtonStyle = styled.button`
  color: ${WHITE_PRIMARY};
  font-family: ${POPPINS};
  font-style: thin;
  text-align: center;
  border-radius: 5px;
  cursor: pointer;
  :focus {
    outline: 0;
  }
  &.tiny {
    height: 26px;
    width: 72px;
  }
  &.small {
    height: 32px;
    width: 102px;
  }
  &.standard {
    height: 40px;
    width: 151px;
  }
  &.large {
    height: 54px;
    width: 160px;
  }

  &.primary {
    background-color: ${PURPLE_PRIMARY_BUTTON};
    border-color: transparent;

    :hover {
      background-color: ${PURPLE_PRIMARY_HOVER};
      box-shadow: ${BOX_SHADOW_6};
    }

    :active {
      background-color: ${PURPLE_PRIMARY_ACTIVE};
    }
  }
  &.secondary {
    background-color: ${PINK_SECONDARY_BUTTON};
    border-color: transparent;

    :hover {
      background-color: ${PINK_SECONDARY_HOVER};
      box-shadow: ${BOX_SHADOW_6};
    }

    :active {
      background-color: ${PINK_SECONDARY_ACTIVE};
    }
  }
  &.hollow-light {
    border: 1px solid ${GREY_4};
    color: ${GREY_9};
    :hover {
      border-color: ${GREY_5};
    }
  }
  &.hollow-dark {
    border: 1px solid ${WHITE_PRIMARY};
    color: ${WHITE_PRIMARY};
    background: transparent;
    :hover {
      background-color: ${WHITE_PRIMARY};
      color: ${PURPLE_DARK};
    }
  }
`;

const SpanStyle = styled.div`
  letter-spacing: 0;
  text-align: center;
  &.tiny {
    font-size: 11px;
    font-weight: 500;
  }
  &.small {
    font-size: 13px;
  }
  &.standard {
    font-size: 14px;
  }
  &.large {
    font-size: 16px;
  }
`;

export default Button;
