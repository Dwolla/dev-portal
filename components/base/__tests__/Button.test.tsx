import renderer from "react-test-renderer";
import Button from "../Button";

test("Button", () => {
  const tree = renderer
    .create(<Button text="Button" size="standard" variant="primary" />)
    .toJSON();

  expect(tree).toMatchSnapshot();
});
