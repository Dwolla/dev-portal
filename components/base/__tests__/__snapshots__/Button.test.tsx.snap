// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button 1`] = `
<button
  className="primary standard css-1mbgqrf-StyledButton-buttonStyles-buttonStyles e12ysvgk0"
  type="button"
>
  Button
</button>
`;

exports[`Link Button 1`] = `
<a
  className="primary standard css-1e7u6fh-StyledLink-buttonStyles-buttonStyles e12ysvgk1"
  href="https://www.dwolla.com"
  onClick={[Function]}
  onMouseEnter={[Function]}
  target="_blank"
>
  Button
</a>
`;
